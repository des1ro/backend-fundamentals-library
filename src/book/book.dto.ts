export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
