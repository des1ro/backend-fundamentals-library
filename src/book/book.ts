export class Book {
  constructor(
    readonly title: string,
    readonly author: string,
    readonly year: number
  ) {}
}
/**
 * book ->
 *  book.service.ts
 *  book.dto.ts
 *
 *
 * book ->
 *    services ->
 *      book.service.ts
 *      newspaper.service.ts
 *    dto ->
 *      book.dto.ts
 *       .....
 */
