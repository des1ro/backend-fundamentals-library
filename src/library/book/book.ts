export class Book {
  constructor(
    private readonly title: string,
    private readonly author: string,
    private readonly year: number,
    private readonly id: number
  ) {}
  getId(): number {
    return this.id;
  }
  getTitle(): string {
    return this.title;
  }
  getAuthor(): string {
    return this.author;
  }
  getYear(): number {
    return this.year;
  }
}
