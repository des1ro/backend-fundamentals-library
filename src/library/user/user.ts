import { randomUUID } from "crypto";

export class User {
  private readonly uuid: string = randomUUID();
  private credit: number = 0;
  constructor(private readonly name: string) {}
  getName(): string {
    return this.name;
  }
  getUuid(): string {
    return this.uuid;
  }
  resetCredit() {
    this.credit = 0; // Czy tu powinienem wrzucić credit do construktora, (private readonly credit) i chcąc zmiany tworzyć new User?
  }
  getCredit() {
    return this.credit;
  }
  setCredit(credit: number): void {
    this.credit = credit;
  }
}
