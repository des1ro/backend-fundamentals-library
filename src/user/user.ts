import { randomUUID } from "crypto";
import { isAfter } from "date-fns";
export class User {
  readonly uuid: string = randomUUID();
  private dueDate?: number;
  private penaltyPoints = 0;
  constructor(readonly name: string) {}
  getPenaltyPoints() {
    return this.penaltyPoints;
  }
  addPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints += penaltyPoints;
  }
  subtractPenaltyPoints(penaltyPoints: number): void {
    this.penaltyPoints -= penaltyPoints;
  }
  getDueDate() {
    return this.dueDate || 0;
  }
  setDueDate(date: number): void {
    this.dueDate = date;
  }
  isBanned(): boolean {
    return isAfter(this.getDueDate(), new Date());
  }
}
