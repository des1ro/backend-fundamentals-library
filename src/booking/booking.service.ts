import { isAfter, addMonths } from "date-fns";
import { Book } from "../book/book.dto";
import { LibraryError } from "../exceptions/library.exceptions";
import { User } from "../user/user.dto";

export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {}
  addBookToUser(book: Book): void {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    this.booksBorrowDate.delete(returnedBook);
  }
  getBooksDatesArray(): Date[] {
    return Array.from(this.booksBorrowDate.values());
  }
}
