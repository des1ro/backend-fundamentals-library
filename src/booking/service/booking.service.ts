import { addMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { LibraryError } from "../../exceptions/library.exceptions";
import { User } from "../../user/user";
export class Booking {
  private readonly penaltyPointsPerBook: number;
  constructor(
    private readonly user: User,
    private readonly booksBorrowDate = new Map<Book, Date>()
  ) {
    this.penaltyPointsPerBook = 10;
  }
  addBookToUser(book: Book): void {
    if (this.user.isBanned()) {
      throw new LibraryError({
        name: "PERMISION_DENIED",
        message: "User is banned!",
      });
    }
    this.booksBorrowDate.set(book, new Date());
  }
  returnBook(returnedBook: Book): void {
    const date = this.booksBorrowDate.get(returnedBook);
    if (!date) {
      throw new LibraryError({
        name: "INVALID_BOOK",
        message: "The book is not in the user's list",
      });
    }
    const bookBorrowDateLimit = addMonths(date, 1);
    const bookIsAfterDueDate = isAfter(new Date(), bookBorrowDateLimit);
    this.booksBorrowDate.delete(returnedBook);
    if (bookIsAfterDueDate) {
      this.user.subtractPenaltyPoints(this.penaltyPointsPerBook);
    }
  }
  has(book: Book): boolean {
    return this.booksBorrowDate.has(book);
  }
  _getBooksDatesArray(): Date[] {
    const datesArray = Array.from(this.booksBorrowDate.values());
    return datesArray;
  }
}
