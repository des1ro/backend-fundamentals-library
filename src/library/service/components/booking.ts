import { add, addMonths, differenceInMonths, isAfter } from "date-fns";
import { Book } from "../../book/book";
import { User } from "../../user/user";
import { LibraryError } from "../../exceptions/library.exceptions";
// sprawdzać przy oddawaniu czy książka po terminie
// wszystko w bookingu
// user ma mieć
export class Booking {
  constructor(
    private readonly user: User,
    private readonly booksDueDate = new Map<Book, Date>()
  ) {}
  getUser(): User {
    return this.user;
  }
  isUserBanned(
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): boolean {
    const isBanned = isAfter(this.user.getDueDate(), new Date());
    if (isBanned) {
      return true;
    }
    const userPenaltyPoints = this.checkUserPenaltyPoints(
      bookBorrowMonthsLimit,
      penaltyPointsLimit
    );
    if (userPenaltyPoints > 0) {
      const banMonths = userPenaltyPoints / 10;
      this.user.setDueDate(addMonths(new Date(), banMonths));
      return true;
    }
    return false;
  }
  addBookToUser(
    book: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsLimit: number
  ): void {
    if (!this.isUserBanned(bookBorrowMonthsLimit, penaltyPointsLimit)) {
      this.booksDueDate.set(book, new Date());
      return;
    }
    throw new LibraryError({
      name: "PERRMISION_DENIED",
      message: "User is banned!",
    });
  }
  checkUserPenaltyPoints(
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): number {
    let penaltyPoints = 0;
    this.booksDueDate.forEach((bookDate) => {
      const difference = differenceInMonths(new Date(), bookDate);
      if (difference >= bookBorrowMonthsLimit)
        penaltyPoints += penaltyPointsPerBook;
    });
    this.user.addPenaltyPoints(penaltyPoints);
    return this.user.getPenaltyPoints();
  }

  returnBook(
    returnedBook: Book,
    bookBorrowMonthsLimit: number,
    penaltyPointsPerBook: number
  ): void {
    const isUserBanned = this.isUserBanned(
      bookBorrowMonthsLimit,
      penaltyPointsPerBook
    );
    if (this.booksDueDate.has(returnedBook)) {
      const date = this.booksDueDate.get(returnedBook);
      if (date && isAfter(new Date(), addMonths(date, 1)) && isUserBanned) {
        this.user.addPenaltyPoints(-penaltyPointsPerBook);
      }
      this.booksDueDate.delete(returnedBook);
      return;
    }
    throw new LibraryError({
      name: "INVALID_BOOK",
      message: "The book is not in the user's list",
    });
  }
  getBooksDueDate(): Map<Book, Date> {
    return this.booksDueDate;
  }
  getUserDueDate() {
    return this.user.getDueDate();
  }
  getUserPenaltyPoints() {
    return this.user.getPenaltyPoints();
  }
}
