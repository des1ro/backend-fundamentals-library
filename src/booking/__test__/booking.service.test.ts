import { randomUUID } from "crypto";
import { Book } from "../../book/book.dto";
import { User } from "../../user/user.dto";
import { Booking } from "../booking.service";
import { addMonths } from "date-fns";
import { LibraryError } from "../../exceptions/library.exceptions";
jest.mock("crypto", () => ({
  randomUUID: jest.fn().mockReturnValue("test-uuid"),
}));
describe("Booking test suite", () => {
  let objectUnderTest: Booking;
  let mockedUser: User;
  let mockedBookOne: Book;
  let mockedBookTwo: Book;
  let mockedDate: number;
  let testBookBorrowMonthsLimit: number;
  let testPenaltyPointsPerBook: number;
  beforeAll(() => {
    mockedBookOne = new Book("Test title One", "Test author One", 1999);
    mockedBookTwo = new Book("Test title Two", "Test author Two", 1995);
    testBookBorrowMonthsLimit = 1;
    testPenaltyPointsPerBook = 10;
  });
  beforeEach(() => {
    mockedUser = new User("Test user");
    objectUnderTest = new Booking(mockedUser);

    jest.useFakeTimers();
  });
  afterEach(() => {
    jest.useRealTimers();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  describe("Return book suite test", () => {
    it("Should add book to user correctly", () => {
      //Given
      jest.setSystemTime(new Date(2023, 0, 20, 12, 30, 0, 0));
      //When
      objectUnderTest.addBookToUser(mockedBookOne);
      const result = objectUnderTest;
      //Then
      expect(result).toMatchSnapshot();
    });
    it("Should throw library error if banned user want to borrow", () => {
      //Given
      mockedDate = addMonths(new Date(), 1).getTime();
      const spy = jest.spyOn(mockedUser, "getDueDate");
      //When
      spy.mockReturnValue(mockedDate);
      //Then
      expect(() => objectUnderTest.addBookToUser(mockedBookOne)).toThrow(
        LibraryError
      );
      expect(spy).toBeCalledTimes(1);
    });
  });

  describe("Return book suite test", () => {
    it("Should throw error if user want to return wrong book", () => {
      //Given
      objectUnderTest.addBookToUser(mockedBookOne);
      //When
      const spy = jest.spyOn(objectUnderTest, "returnBook");
      //Then
      expect(() => objectUnderTest.returnBook(mockedBookTwo)).toThrow(
        LibraryError
      );
      expect(spy).toBeCalledTimes(1);
    });
    it("Should return book correctly", () => {
      //Given
      const spy = jest.spyOn(objectUnderTest, "returnBook");
      //When
      objectUnderTest.addBookToUser(mockedBookOne);
      //Then
      expect(() => objectUnderTest.returnBook(mockedBookOne)).not.toThrow();
      expect(objectUnderTest).toMatchSnapshot();
      expect(spy).toBeCalledTimes(1);
    });
  });
});
