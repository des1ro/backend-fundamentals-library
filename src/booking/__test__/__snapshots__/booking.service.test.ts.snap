// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Booking test suite Return book suite test Should add book to user correctly 1`] = `
Booking {
  "booksBorrowDate": Map {
    Book {
      "author": "Test author One",
      "title": "Test title One",
      "year": 1999,
    } => 2023-01-20T11:30:00.000Z,
  },
  "user": User {
    "dueDate": 0,
    "name": "Test user",
    "penaltyPoints": 0,
    "uuid": "test-uuid",
  },
}
`;

exports[`Booking test suite Return book suite test Should return book correctly 1`] = `
Booking {
  "booksBorrowDate": Map {},
  "returnBook": [MockFunction] {
    "calls": [
      [
        Book {
          "author": "Test author One",
          "title": "Test title One",
          "year": 1999,
        },
      ],
    ],
    "results": [
      {
        "type": "return",
        "value": undefined,
      },
    ],
  },
  "user": User {
    "dueDate": 0,
    "name": "Test user",
    "penaltyPoints": 0,
    "uuid": "test-uuid",
  },
}
`;
